src/fetch_events.py

```diff
@@ -1,133 +1,81 @@
+    ]
+    object_summaries = itertools.chain(*object_summaries_by_timestamp)
 
-    command = f"aws s3 sync {uri_s3} {path_local}"
+    # Spread fetching tasks over a number of CPU threads. Until aioboto3 is thoroughly documented and isn't a pain
+    # to work with (or until boto3 is asyncio-friendly), multithreading is a decent alternative
+    # Fetching all data (even gzipped) from the get-go might incur significant memory footprint, but this is a simple start
+    with ThreadPoolExecutor(max_workers=num_concurrent_downloads) as executor:
+        dfs = executor.map(_fetch_decompress_parse, object_summaries)
 
-    # TODO: Once a global logging config is set, log the exact command to the console
-    # Would also be helpful to customize log message by adding file name, function & line
-    # number of where the log happens, kind of like this:
-    # https://github.com/LocalAtBrown/experiment-semantic-similarity/blob/main/notebooks/util/logging.py
+    return pd.concat(dfs)
 
-    # Run the command
-    return subprocess.Popen(command.split(" "), stdout=subprocess.DEVNULL, stderr=subprocess.PIPE)
 
+def _fetch_decompress_parse(object_summary: ObjectSummary) -> pd.DataFrame:
+    key, response = _fetch_object(object_summary)
+    data = _decompress_object(key, response)
+    df = _parse_object(key, data)
+    return df
 
-def _read_folder(site_bucket_name: str, timestamp: datetime, path_to_data_dir: Path) -> pd.DataFrame:
-    """
-    Opens all .gz files in a local data folder given its corresponding site bucket
-    name and date-hour timestamp and combine all their data into a single pandas DataFrame.
-    """
-    # Local directory holding all data files fetched from the remote S3 folder
-    path_local = path_to_data_dir / convert_to_s3_folder(timestamp)
 
-    # Probably only need *.gz instead of **/*.gz because there should be no directory
-    # within the local data folder, but using the latter for good measure
-    dfs = [read_gzipped_json_records(path_file) for path_file in path_local.glob("**/*.gz")]
+def _fetch_object(object_summary: ObjectSummary) -> Tuple[str, GetObjectOutputTypeDef]:
+    key = object_summary.key
+    try:
+        response = object_summary.get()
+    except Exception as e:
+        raise EventObjectFetchError(f"Failed to fetch object {object_summary.key}") from e
+    return key, response
 
-    # If there are no files/DataFrames, return an empty DataFrame, since pd.concat([]) throws an error
-    if len(dfs) == 0:
-        # TODO: logging.info showing no data available to fetch for this particular date-hour timestamp
-        return pd.DataFrame()
 
-    # Combine into a single DataFrame
-    return pd.concat(dfs)
+def _decompress_object(object_key: str, object_response: GetObjectOutputTypeDef) -> bytes:
+    try:
+        data = object_response["Body"].read()
+        data = gzip.decompress(data)
+    except Exception as e:
```

Comment:
❓ is there a more specific exception we can `except`?

Reply:
This try-except block and the one in `_parse_object` are a bit more tricky to anticipate. Maybe we can leave them as-is and see what happens?

Reply:
Happy to be wrong, though!

Reply:
Works for me